teleport_tube.lua

```diff
@@ -1,206 +1,304 @@
-local S = minetest.get_translator("pipeworks")
-local filename=minetest.get_worldpath() .. "/teleport_tubes" -- Only used for backward-compat
-local storage=minetest.get_mod_storage()
-
-local tp_tube_db = nil -- nil forces a read
-local tp_tube_db_version = 4
 
--- cached rceiver list: hash(pos) => {receivers}
-local cache = {}
-
-local function hash(pos)
-	return string.format("%.30g", minetest.hash_node_position(pos))
+local S = minetest.get_translator("pipeworks")
+local filename = minetest.get_worldpath().."/teleport_tubes"  -- Only used for backward-compat
+local storage = minetest.get_mod_storage()
+
+local has_digilines = minetest.get_modpath("digilines")
+
+-- V1: Serialized text file indexed by vector position.
+-- V2: Serialized text file indexed by hash position.
+-- V3: Mod storage using serialized tables.
+-- V4: Mod storage using "<can_receive>:<channel>" format.
+local tube_db_version = 4
+local tube_db = {}
+local receiver_cache = {}
+
+local function hash_pos(pos)
+	vector.round(pos)
+	return string.format("%.0f", minetest.hash_node_position(pos))
 end
 
--- New entry format: <can receive (number)> ':' <channel (string)>
-
-local function serialize_tube_db_entry(entry)
-	return entry.cr .. ":" .. entry.channel
+local function serialize_tube(tube)
+	return string.format("%d:%s", tube.cr, tube.channel)
 end
 
-local function deserialize_tube_db_entry(hash, str)
-	local sep = str:find(":", 2, true)
-	if not sep then return minetest.deserialize(str) end
-	local cr = tonumber(str:sub(1, sep - 1))
-	if not cr then return minetest.deserialize(str) end
-	local channel = str:sub(sep + 1)
-	local pos = minetest.get_position_from_hash(tonumber(hash))
-	return {x = pos.x, y = pos.y, z = pos.z, cr = cr, channel = channel}
+local function deserialize_tube(hash, str)
+	local sep = str:sub(2, 2) == ":"
+	local cr = tonumber(str:sub(1, 1))
+	local channel = str:sub(3)
+	if sep and cr and channel then
+		local pos = minetest.get_position_from_hash(tonumber(hash))
+		return {x = pos.x, y = pos.y, z = pos.z, cr = cr, channel = channel}
+	end
 end
```

Comment:
i'm not intimate w/ this code, but these don't seem equivalent at all. is "cr" *always* a single digit? why not do something like
```lua
local cr, channel = unpack(str:split(":", true, 1))
cr = tonumber(cr)
if cr and channel then
    local pos = minetest.get_position_from_hash(tonumber(hash))
    pos.cr = cr
    pos.channel = channel
    return pos
else
    return minetest.deserialize(str)
end
```

Reply:
oh i see it's "can receive" and is either 0 or 1? so i guess it's always a single number.

Reply:
and the de-serialize stuff is handled in the update code. got it.

Reply:
Yes, `cr` (short for `can_receive`) is always either 0 or 1, so it's much more efficient to use `string.sub`.